### PR TITLE
🏗 Parallelizes execution of unit tests and e2e test on CircleCI

### PR DESCRIPTION
**Pull request by danielrozenberg**
This PR:
- Parallelizes unit tests and e2e test (for both the base flavor and experimental flavors) into 16 shards, using CircleCI's [build-in support](https://circleci.com/docs/2.0/parallelism-faster-jobs/) for parallelizing tests
  - Parallelized jobs now have the `⛓️` emoji in their name to indicate that they're paralellized
- Splits the `--local_changes` into a separate job which is a prerequisite to the (now renamed) `All Unit Tests` job
  - This job explicitly halts early and passes execution to `All Unit Tests` when running on push builds, as it has no meaning outside of pull requests
- Splits experiment builds' e2e tests and integration tests into separate jobs to enable parallelization on the e2e tests

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -352,9 +396,11 @@ workflows:
             - 'Nomodule Build'
             - 'Module Build'
       - 'End-to-End Tests':
+          name: '⛓️ End-to-End Tests'
           <<: *push_and_pr_builds
           requires:
             - 'Nomodule Build'
+            - '⛓️ All Unit Tests'
```
Comment (rsimha) [inline .circleci/config.yml:403]: Was it your intent to make the E2E tests wait for the Unit tests to reduce VM usage? Instead, WDYT about getting rid of this dependency, and playing with the overall parallelism number (see other comment)?

File: .circleci/config.yml [hunk 2]
```diff
@@ -178,18 +178,39 @@ jobs:
...
-          command: node build-system/pr-check/unit-tests.js
+          name: 'Local Unit Tests'
+          command: node build-system/pr-check/unit-tests-local.js
       - store_test_output
       - fail_fast
       - teardown_vm
+  'All Unit Tests':
+    executor:
+      name: amphtml-large-executor
+    parallelism: 16
```
Comment (rsimha) [inline .circleci/config.yml:205]: CircleCI gives our entire org a max of ~80 parallel VMs, after which it starts queueing. Do we know if this number counts towards that quota? (I'm wondering if this PR could result in more queueing once it is merged, even if the builds run faster.)

Can we experiment with lower parallelism numbers (say, 4 and 8) and see what the effects are on overall CPU consumption and running time? (E.g. each parallel shard appears to be doing things like checking out code, installing packages, installing chrome, etc.)

I'm guessing the sweet spot lies below 16, but it would be nice to back that up with real results, and reduce any potential cause for increased queueing of jobs.

**Review by rsimha — COMMENTED** [review]
Excited to see the results of this work. A few initial comments below, hopefully all useful :)

File: .circleci/config.yml [hunk 3]
```diff
@@ -178,18 +178,39 @@ jobs:
...
     steps:
+      - unless:
+          condition:
+            matches:
+              pattern: ^pull/\d+$
+              value: << pipeline.git.branch >>
+          steps:
+            - run:
+                name: 'Skip Local Unit Test on Push Builds'
+                command: circleci-agent step halt
```
[the remark this answers is not in the document]
Comment (danielrozenberg) [reply .circleci/config.yml:193]: The format here is a bit different - this is a _step_, not a _job_ that needs to be conditioned. I tried setting the job to skip entirely, but then the dependent job (`All Unit Tests`) never executes. As a middle-group, I moved this block to the `jobs:` section - does this look good for you?

File: .circleci/config.yml [hunk 4]
```diff
@@ -275,7 +296,7 @@ jobs:
           command: node build-system/pr-check/experiment-build.js --experiment=experiment<< parameters.exp >>
       - fail_fast
       - teardown_vm
-  'Experiment Tests':
+  'Experiment Integration Tests':
```
[the remark this answers is not in the document]
Comment (danielrozenberg) [reply .circleci/config.yml:299]: Done.

File: build-system/pr-check/experiment-e2e-tests.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+/**
+ * Steps to run during PR builds.
+ */
+function prBuildWorkflow() {
+  if (
+    buildTargetsInclude(
+      Targets.RUNTIME,
+      Targets.INTEGRATION_TEST,
```
[the remark this answers is not in the document]
Comment (danielrozenberg) [reply build-system/pr-check/experiment-e2e-tests.js:76]: Done.

File: build-system/pr-check/experiment-integration-tests.js [hunk 6]
```diff
@@ -44,9 +44,6 @@ function runExperimentTests(config) {
     timedExecOrThrow(
       `amp integration --nobuild --compiled --headless ${experimentFlag} ${defineFlag} ${reportFlag}`
     );
-    timedExecOrThrow(
-      `amp e2e --nobuild --compiled --headless ${experimentFlag} ${defineFlag} ${reportFlag}`
-    );
```
[the remark this answers is not in the document]
Comment (danielrozenberg) [reply build-system/pr-check/experiment-integration-tests.js:49]: Done.

File: build-system/pr-check/experiment-e2e-tests.js [hunk 7]
```diff
@@ -0,0 +1,89 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (danielrozenberg) [reply build-system/pr-check/experiment-e2e-tests.js:2]: Done. I renamed the original file into this, which in some technical way means we shouldn't change it, but it's an edge case...

File: build-system/pr-check/ci-job.js [hunk 8]
```diff
@@ -57,6 +58,22 @@ async function runCiJob(jobName, pushBuildWorkflow, prBuildWorkflow) {
...
 }
 
+/**
+ * Returns list of test files that CircleCI should execute in a parallelized job shard.
+ *
+ * @param {!Array<string>} globs array of glob strings for finding test file paths.
+ * @return {string} a comma separated list of test file paths.
+ */
+function getCircleCiShardTestFiles(globs) {
+  const joinedGlobs = `"${globs.join('" "')}"`;
```
[the remark this answers is not in the document]
Comment (danielrozenberg) [reply build-system/pr-check/ci-job.js:68]: Done, copied your commit to avoid merge conflicts with the other fixes

File: .circleci/config.yml (hunk 1 above)
Comment (danielrozenberg) [reply .circleci/config.yml:403]: Did you read the explanation above about this dependency and why I think it's okay?

btw, I'm suspecting that parallelized jobs _do not_ use an extra VM - the documentation doesn't say anything about it, but my hint is that each job has the `🗳 1 / 83 parallel jobs ℹ️` indicator, and multiple parallelized jobs in my workflow have been running at the same time, saying `🗳 16 / 83 parallel jobs ℹ️` - so it looks like they're not constrained by the same 80 VMs limit

File: .circleci/config.yml (hunk 2 above)
Comment (danielrozenberg) [reply .circleci/config.yml:205]: I experimented with multiple values, 16 seems to be the sweet spot - above 16 the improvement in runtime is marginal-to-nonexistent

As for the shared installation steps, I've seen something in the CircleCI documentation that makes me think we can speed those up (not 100% sure, but definitely want to experiment with it soon)

File: build-system/pr-check/ci-job.js [hunk 9]
```diff
@@ -57,6 +58,22 @@ async function runCiJob(jobName, pushBuildWorkflow, prBuildWorkflow) {
   }
 }
 
+/**
+ * Returns list of test files that CircleCI should execute in a parallelized job shard.
+ *
+ * @param {!Array<string>} globs array of glob strings for finding test file paths.
+ * @return {string} a comma separated list of test file paths.
+ */
+function getCircleCiShardTestFiles(globs) {
```
Comment (rsimha) [inline build-system/pr-check/ci-job.js:67]: This is clever, thanks for figuring it out. I have two high-level comments about this function:
1. The intent behind `ci-job.js` was to be a minimal skeleton that's common across all CI jobs. Since this is a helper that's used by some jobs but not others, a better place for it would be a helper file like `build-system/pr-check/utils.js`.
2. As discussed yesterday, `--files` is mainly used for local development, and its implementation differs from what we use during CI (logging, error handling, choice of reporters, etc.). Also, it has the downside of [very long logging messages](https://user-images.githubusercontent.com/26553114/126356886-d2984bf1-085a-487b-972e-e171eff96b85.png) that make the logs hard to read. I suggest switching this function to write the list of test suites to disk, and adding a new flag called `--filelist` that reads the file passed in. (This can be done by adding a new function `getFilesFromFilelist()` to `build-system/common/utils.js`, similar to `getFilesFromArgv()` that reads the list).

WDYT?

**Review by rsimha — COMMENTED** [review]
Wrote this comment yesterday, but it didn't get published.

Will review the other changes this afternoon.

File: build-system/pr-check/ci-job.js (hunk 9 above)
Comment (danielrozenberg) [reply build-system/pr-check/ci-job.js]: Added commit 20de51b to address (1) and 4c1b552 to address (2), PTAL

Comment (danielrozenberg) [reply build-system/pr-check/ci-job.js]: Also wdyt about `--from_file` for the flag name? (reaching  because this is expected to only be used in CI

Comment (danielrozenberg) [reply build-system/pr-check/ci-job.js]: Also wdyt about --from_file for the flag name? (reaching 🚲🏚 moment because this is expected to only be used in CI)

File: extensions/amp-riddle-quiz/0.1/test/test-amp-riddle-quiz.js [hunk 10]
```diff
@@ -32,7 +32,8 @@ describes.realWin(
       win.document.body.appendChild(element);
     });
 
-    it('should have iframe when built', () => {
+    // TODO(#): disabled in #35227
+    it.skip('should have iframe when built', () => {
```
Comment (rileyajones) [inline extensions/amp-riddle-quiz/0.1/test/test-amp-riddle-quiz.js:36]: It would be nice to file a ticket and associate these `TODO`s with it.

Comment (danielrozenberg) [reply extensions/amp-riddle-quiz/0.1/test/test-amp-riddle-quiz.js:36]: It's on the PR's TODOs :)

File: .circleci/config.yml [hunk 11]
```diff
@@ -86,6 +86,17 @@ commands:
           path: result-reports
       - store_test_results:
           path: result-reports
+  skip_on_push_builds:
+    steps:
+      - unless:
+          condition:
+            matches:
+              pattern: ^pull/\d+$
```
Comment (rsimha) [inline .circleci/config.yml:94]: This pattern won't work in all cases because PR branches from forks start with `pull/`, while PR branches on the main repo can have arbitrary names like `danielrozenberg-patch-1`. I've suggested an alternative below.

File: .circleci/config.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (rsimha) [reply .circleci/config.yml]: To clarify, I was suggesting that the condition implementation be added at the top as follows:

```yaml
pr_builds_only: &pr_builds_only
  filters:
    branches:
      ignore:
        - main
        - /^amp-release-.*$/
```

Then, it can be used by a step as follows:

```yaml
  command_name:
    steps:
      <<: *pr_builds_only
      - run:
          command: <command to run during pr builds>
```

This way, it can also be reused in future by other jobs / steps. (And if this particular step needs to run for just push builds, you could similarly reuse the `push_builds_only` condition.)

[the remark this answers is not in the document]
Comment (danielrozenberg) [reply .circleci/config.yml]: This isn't a `filter` though - a `filter` controls whether an entire _job_ runs or not, while this particular case requires that one individual _step_ be run - I don't want to create `pr_builds_only` defined as a step-controller, because `push_builds_only` is a job-controller - so there's a subtle inconsistency here that could hurt us down the line

File: build-system/pr-check/experiment-integration-tests.js [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+const {experiment} = require('minimist')(process.argv.slice(2));
+const {getExperimentConfig} = require('../common/utils');
+const {isPushBuild} = require('../common/ci');
+const {runCiJob} = require('./ci-job');
+const {Targets, buildTargetsInclude} = require('./build-targets');
+
+const jobName = `${experiment}-integration-tests.js`;
+
+/**
+ * Runs tests for the given configuration and reports results for push builds.
```
Comment (rsimha) [inline build-system/pr-check/experiment-integration-tests.js:36]: nit: This is executed during PR and push builds.
```suggestion
 * Runs tests for the given configuration and reports results.
```

File: .circleci/config.yml (hunk 11 above)
Comment (danielrozenberg) [reply .circleci/config.yml]: Done.

File: build-system/pr-check/ci-job.js (hunk 9 above)
Comment (rsimha) [reply build-system/pr-check/ci-job.js]: Thanks for making this change. I think the flag ought to be `--filelist` (not `--fileslist`) because it is a well-known standard concept / name. (E.g. Google search [redirects](https://www.google.com/search?q=FilesList) `FilesList` to `FileList` 😃 )

Comment (danielrozenberg) [reply build-system/pr-check/ci-job.js]: Done.

File: .circleci/config.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (rsimha) [reply .circleci/config.yml]: Your point about filters has convinced me. Okay to leave this as is 👍

File: .circleci/config.yml (hunk 1 above)
Comment (danielrozenberg) [reply .circleci/config.yml:406]: Now that I'm re-reading my own explanations, I realize that this change only makes sense if the assumption in my previous comment on this thread is wrong - which is not what the signs point to. Reverted the change.

File: .circleci/config.yml (hunk 2 above)
Comment (danielrozenberg) [reply .circleci/config.yml:208]: It's a bit difficult to gather because I didn't keep the data around, but here are some samples:

- End-to-End tests with 8 shards: [5m27s](https://app.circleci.com/pipelines/github/ampproject/amphtml/12938/workflows/0c0118ef-9d3d-4ac7-8b26-2adf17bab69d/jobs/207377)
- Unit Tests with 4 shards: [more than 8m55s](https://app.circleci.com/pipelines/github/ampproject/amphtml/12897/workflows/307caaa1-02e6-4af3-9cb5-8e1f785c2ff1/jobs/206565) ("more" because `fail_fast` stopped these before they completed)
- Unit Tests with 8 shards: [more than 3m35s](https://app.circleci.com/pipelines/github/ampproject/amphtml/12926/workflows/ff311539-6fc2-4d41-9f76-1677d727b596/jobs/207111)
- Unit Tests with 32 shards: [4m31s](https://app.circleci.com/pipelines/github/ampproject/amphtml/12928/workflows/20a9e7fb-40b6-4e0d-a0e2-e2c7c9f229ab/jobs/207161)

With 16 shards for each, they take between 3m and 3m30s. Admittedly this isn't super scientific, but I think this is Good Enough™ for now

Comment (rsimha) [reply .circleci/config.yml:208]: I compared recent [`main` branch builds](https://app.circleci.com/insights/github/ampproject/amphtml/workflows/CircleCI/overview?branch=main&reporting-window=last-7-days) against [builds for this PR branch](https://app.circleci.com/insights/github/ampproject/amphtml/workflows/CircleCI/overview?branch=pull%2F35227&reporting-window=last-7-days):

`main`: ~2200 credits for ~16 minute builds.

image


This PR: ~4300 credits for ~14 minute builds.

image

By increasing parallelism to 16, it appears we're spending nearly 100% more $$ for a < 15% speed up.
- Do these numbers seem right to you?
- Can we retry with parallelism factors of 4 and 8 to see if there's a better sweet spot for total duration vs. total credit usage for an average build (that affects the AMP runtime)?

**Review by rileyajones — APPROVED** [review]

File: .circleci/config.yml (hunk 2 above)
Comment (danielrozenberg) [reply .circleci/config.yml:208]: Ah! So they _do_ charge extra for the parallelization. ok, let's try a few executions with 8 / 4 parallels

Comment (rsimha) [reply .circleci/config.yml]: Yep. And while you're at it, can we generate a build with 2 parallel shards as well? After that, once they all show up on the insights dashboard for this PR, we can compare costs and benefits for 2 through 16. (These numbers don't have to all be powers of 2 😃 )

**Comment by danielrozenberg** [comment]
I experimented with parallelism numbers and Docker container sizes for the newly-parallelized jobs, and collected the data here: https://docs.google.com/spreadsheets/d/11JOAxtNTCzV7KxKXAwVrHhjHvsoVBW-JKEmsTL8BK3U/edit#gid=1450490992

Given the averages, setting parallelism to 6 and container size to `medium` seems like a good choice - it increases the credits used from (average of) 2,138 to 2,388 but reduces the (average) total duration from 18:09 minutes to 13:58 minutes, which is a huge improvement at a marginal cost.

With followup PRs I believe I can lower the credits used again

File: .circleci/config.yml (hunk 2 above)
Comment (danielrozenberg) [reply .circleci/config.yml]: I experimented with parallelism numbers and Docker container sizes for the newly-parallelized jobs, and collected the data here: https://docs.google.com/spreadsheets/d/11JOAxtNTCzV7KxKXAwVrHhjHvsoVBW-JKEmsTL8BK3U/edit#gid=1450490992

Given the averages, setting parallelism to 6 and container size to medium seems like a good choice - it increases the credits used from (average of) 2,138 to 2,388 but reduces the (average) total duration from 18:09 minutes to 13:58 minutes, which is a huge improvement at a marginal cost.

With followup PRs I believe I can lower the credits used again

Comment (rsimha) [reply .circleci/config.yml]: This is brilliant! Thanks for going through this exercise. I agree with your conclusion.

**Review by rsimha — APPROVED** [review]
This was a valiant effort, and the results in https://github.com/ampproject/amphtml/pull/35227#discussion_r675703617 are very exciting. A ~30% reduction in running time for a ~10% increase in credits used is an excellent tradeoff.

LGTM!
